types/jquery/JQuery.d.ts

```diff
@@ -83,6 +83,7 @@ $( document.body )
      * @since 1.4
      */
     add(selector: JQuery.Selector, context: Element): this;
+    // TODO: The return type should reflect newly selected types.
```

Comment:
I am not sure I really understand this TODO comment. How can we represent the element types after added?

Reply:
```typescript
const paragraphs = document.querySelectorAll('p');
// Array.from strictly isn't necessary but there appears to be a bug in the declarations. Passing an ArrayLike should work.
const $p = $(Array.from(paragraphs)); // JQuery<HTMLParagraphElement>
const $a = $p.add('div');
```

With the current declarations, `$a` will be of type `JQuery<HTMLParagraphElement>` even though it may include `HTMLDivElement` objects. `$a` should probably be of type `JQuery<HTMLParagraphElement | HTMLDivElement>`. This can be accomplished by changing the return type of `.add()` from `this` to `JQuery<TElement | TAddedType>` where `TAddedType` would depend on the argument passed to `.add()`.

I didn't make the change in this PR for several reasons.
* I wasn't sure on the impact to the user experience. Would this change make certain scenarios unnecessarily difficult or awkward to use?
* This isn't the only API affected by this. The jQuery API should be audited to determine which APIs are affected.
* I haven't determined a cohesive, consistent solution. The solution needs to consider when it's appropriate to return a more specific type, what type to return when it's not possible to infer it, what to do for mixed-DOM scenarios (e.g. HTML + SVG), etc.

Reply:
How about `$p.add('.className')`? What is the returning type?

Reply:
Since selecting by class name could select both HTML and SVG elements, the return type _should_ be `JQuery<HTMLParagraphElement | Element>`.

However, in the current declarations there exists a call signature
```typescript
<TElement extends Element = HTMLElement>(selector: JQuery.Selector, context?: Element | Document | JQuery): JQuery<TElement>;
```
that returns `HTMLElement` instead of `Element`. This was done for usability (vast majority of users will only use jQuery with HTML elements) and backwards compatibility (`@types/jquery@1` and formerly `@types/jquery@2` use `HTMLElement` as their selected type).

So for correctness, `.add('.className')` should return `Element` but for consistency, it should return `HTMLElement`.